File: Harvest/Services/MatchService.swift
import Foundation
import Supabase

struct MatchService {
    private var client: SupabaseClient { SupabaseManager.shared.client }
    private let chatService = ChatService()
    private let profileService = ProfileService()

    func getMatches(userId: String) async throws -> [MatchWithProfile] {
        let matches: [Match] = try await client
            .from("matches")
            .select()
            .or("user1_id.eq.\(userId),user2_id.eq.\(userId)")
            .eq("is_active", value: true)
            .order("matched_at", ascending: false)
            .execute()
            .value

        var matchesWithProfiles: [MatchWithProfile] = []
        var seenOtherUserIds = Set<String>()
        for match in matches {
            guard let otherUserId = match.otherUserId(currentUserId: userId) else { continue }
            guard seenOtherUserIds.insert(otherUserId).inserted else { continue }
            if let profile = try await profileService.getProfile(userId: otherUserId) {
                matchesWithProfiles.append(MatchWithProfile(match: match, profile: profile))
            }
        }

        return matchesWithProfiles
    }

    func unmatchUser(matchId: String) async throws {
        try await client
            .from("matches")
            .update([
                "is_active": AnyJSON.bool(false),
                "unmatched_at": AnyJSON.string(ISO8601DateFormatter().string(from: Date()))
            ])
            .eq("id", value: matchId)
            .execute()
    }

    func blockUser(userId: String, blockedUserId: String) async throws {
        try await client
            .from("user_blocks")
            .insert([
                "blocker_id": userId,
                "blocked_id": blockedUserId
            ])
            .execute()
    }

    func reportUser(reporterId: String, reportedUserId: String, category: String, description: String) async throws {
        try await client
            .from("user_reports")
            .insert([
                "reporter_id": reporterId,
                "reported_id": reportedUserId,
                "reason": category,
                "description": description
            ])
            .execute()
    }

    func ensureConversation(match: Match, currentUserId: String) async throws -> String? {
        try await chatService.ensureConversation(
            matchId: match.id,
            user1Id: match.user1Id,
            user2Id: match.user2Id
        )
    }

    func getMatch(matchId: String) async throws -> Match? {
        let matches: [Match] = try await client
            .from("matches")
            .select()
            .eq("id", value: matchId)
            .limit(1)
            .execute()
            .value

        return matches.first
    }

    func getConversations(userId: String) async throws -> [ConversationWithProfile] {
        let directConversations: [Conversation] = try await client
            .from("conversations")
            .select()
            .or("user1_id.eq.\(userId),user2_id.eq.\(userId)")
            .order("last_message_at", ascending: false)
            .execute()
            .value

        let matches: [Match] = try await client
            .from("matches")
            .select()
            .or("user1_id.eq.\(userId),user2_id.eq.\(userId)")
            .eq("is_active", value: true)
            .execute()
            .value

        let matchIds = matches.map(\.id)
        var conversationsById = Dictionary(uniqueKeysWithValues: directConversations.map { ($0.id, $0) })

        if !matchIds.isEmpty {
            let matchLinkedConversations: [Conversation] = try await client
                .from("conversations")
                .select()
                .in("match_id", values: matchIds)
                .order("last_message_at", ascending: false)
                .execute()
                .value

            for conversation in matchLinkedConversations {
                conversationsById[conversation.id] = conversation
            }
        }

        let matchesById = Dictionary(uniqueKeysWithValues: matches.map { ($0.id, $0) })
        let conversations = conversationsById.values.sorted { lhs, rhs in
            (lhs.lastMessageAt ?? lhs.createdAt ?? "") > (rhs.lastMessageAt ?? rhs.createdAt ?? "")
        }

        var conversationsWithProfiles: [ConversationWithProfile] = []
        var seenConversationIds = Set<String>()
        for conversation in conversations {
            guard seenConversationIds.insert(conversation.id).inserted else { continue }

            let otherUserId =
                conversation.otherUserId(currentUserId: userId) ??
                conversation.matchId.flatMap { matchesById[$0]?.otherUserId(currentUserId: userId) }

            guard let otherUserId else { continue }
            if let profile = try await profileService.getProfile(userId: otherUserId) {
                let hydratedConversation = try await hydrateConversationPreviewIfNeeded(conversation)
                guard hydratedConversation.lastMessagePreview != nil else { continue }
                let hasReplyHighlight = try await shouldHighlightConversation(
                    hydratedConversation,
                    currentUserId: userId
                )
                conversationsWithProfiles.append(
                    ConversationWithProfile(
                        conversation: hydratedConversation,
                        profile: profile,
                        hasReplyHighlight: hasReplyHighlight
                    )
                )
            }
        }

        return conversationsWithProfiles
    }

    func getInboundLikes(userId: String) async throws -> [InboundLikeWithProfile] {
        let inboundSwipes: [Swipe] = try await client
            .from("swipes")
            .select()
            .eq("swiped_id", value: userId)
            .in("action", values: [SwipeAction.like.rawValue, SwipeAction.superLike.rawValue])
            .order("created_at", ascending: false)
            .execute()
            .value

        let matches: [Match] = try await client
            .from("matches")
            .select()
            .or("user1_id.eq.\(userId),user2_id.eq.\(userId)")
            .eq("is_active", value: true)
            .execute()
            .value

        let matchedUserIds = Set(matches.compactMap { $0.otherUserId(currentUserId: userId)?.lowercased() })
        var seenSwiperIds = Set<String>()
        var inboundLikes: [InboundLikeWithProfile] = []

        for swipe in inboundSwipes {
            let swiperId = swipe.swiperId.lowercased()
            guard !matchedUserIds.contains(swiperId) else { continue }
            guard seenSwiperIds.insert(swiperId).inserted else { continue }

            if let profile = try await profileService.getProfile(userId: swipe.swiperId) {
                inboundLikes.append(
                    InboundLikeWithProfile(
                        swipe: swipe,
                        profile: profile
                    )
                )
            }
        }

        return inboundLikes
    }

    private func hydrateConversationPreviewIfNeeded(_ conversation: Conversation) async throws -> Conversation {
        if conversation.lastMessagePreview != nil {
            return conversation
        }

        let latestMessages: [Message] = try await client
            .from("messages")
            .select("content, created_at, id, conversation_id, sender_id, message_type, media_url, is_read, read_at")
            .eq("conversation_id", value: conversation.id)
            .order("created_at", ascending: false)
            .limit(1)
            .execute()
            .value

        guard let latestMessage = latestMessages.first, let content = latestMessage.content, !content.isEmpty else {
            return conversation
        }

        try await client
            .from("conversations")
            .update([
                "last_message_at": AnyJSON.string(latestMessage.createdAt ?? ISO8601DateFormatter().string(from: Date())),
                "last_message_preview": AnyJSON.string(String(content.prefix(100)))
            ])
            .eq("id", value: conversation.id)
            .execute()

        var hydratedConversation = conversation
        hydratedConversation.lastMessageAt = latestMessage.createdAt
        hydratedConversation.lastMessagePreview = String(content.prefix(100))
        return hydratedConversation
    }

    private func shouldHighlightConversation(_ conversation: Conversation, currentUserId: String) async throws -> Bool {
        let latestMessages: [Message] = try await client
            .from("messages")
            .select("content, created_at, id, conversation_id, sender_id, message_type, media_url, is_read, read_at")
            .eq("conversation_id", value: conversation.id)
            .order("created_at", ascending: false)
            .limit(1)
            .execute()
            .value

        guard let latestMessage = latestMessages.first else { return false }
        return !latestMessage.isSentBy(currentUserId)
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/data/service/MatchService.kt
package com.harvestglass.harvest.data.service

import com.harvestglass.harvest.data.model.Conversation
import com.harvestglass.harvest.data.model.ConversationWithProfile
import com.harvestglass.harvest.data.model.InboundLikeWithProfile
import com.harvestglass.harvest.data.model.Match
import com.harvestglass.harvest.data.model.MatchWithProfile
import com.harvestglass.harvest.data.model.Message
import com.harvestglass.harvest.data.model.Swipe
import com.harvestglass.harvest.data.model.SwipeAction
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant

/**
 * The slice of Harvest/Services/MatchService.swift the app still reaches: the
 * conversation list, active matches, inbound likes, and the safety actions.
 *
 * The swipe deck itself stayed behind with Discover — nothing creates a new
 * swipe now except replying to an inbound like.
 */
class MatchService(
    private val client: SupabaseClient,
    private val profileService: ProfileService
) {

    // MARK: - Conversation list

    suspend fun getConversations(userId: String): List<ConversationWithProfile> {
        val blocked = getBlockedUserIds(userId)

        val matches = client.postgrest.from("matches")
            .select {
                filter {
                    or {
                        eq("user1_id", userId)
                        eq("user2_id", userId)
                    }
                    eq("is_active", true)
                }
            }
            .decodeList<Match>()

        if (matches.isEmpty()) return emptyList()

        val matchesById = matches.associateBy { it.id }
        val conversations = client.postgrest.from("conversations")
            .select {
                filter { isIn("match_id", matches.map { it.id }) }
                order("last_message_at", Order.DESCENDING)
            }
            .decodeList<Conversation>()
            .distinctBy { it.id }
            // Newest activity first, falling back to creation time.
            .sortedByDescending { it.lastMessageAt ?: it.createdAt ?: "" }

        val result = mutableListOf<ConversationWithProfile>()
        for (conversation in conversations) {
            val otherUserId = conversation.otherUserId(userId)
                ?: conversation.matchId?.let { matchesById[it]?.otherUserId(userId) }
                ?: continue

            if (blocked.contains(otherUserId.lowercase())) continue

            val profile = profileService.getProfile(otherUserId) ?: continue
            val hydrated = hydrateConversationPreviewIfNeeded(conversation)
            // A conversation with nothing said in it stays out of the list.
            if (hydrated.lastMessagePreview == null) continue

            result += ConversationWithProfile(
                conversation = hydrated,
                profile = profile,
                hasReplyHighlight = shouldHighlightConversation(hydrated, userId)
            )
        }
        return result
    }

    /**
     * Fills in `last_message_preview` for conversations created before the
     * column existed, so they don't silently vanish from the list.
     */
    private suspend fun hydrateConversationPreviewIfNeeded(conversation: Conversation): Conversation {
        if (conversation.lastMessagePreview != null) return conversation

        val latest = latestMessage(conversation.id) ?: return conversation
        val content = latest.content?.takeIf { it.isNotEmpty() } ?: return conversation

        val at = latest.createdAt ?: Instant.now().toString()
        val preview = content.take(PREVIEW_LENGTH)

        client.postgrest.from("conversations").update(
            buildJsonObject {
                put("last_message_at", at)
                put("last_message_preview", preview)
            }
        ) { filter { eq("id", conversation.id) } }

        return conversation.copy(lastMessageAt = at, lastMessagePreview = preview)
    }

    /** True when the newest message came from the other person. */
    private suspend fun shouldHighlightConversation(
        conversation: Conversation,
        currentUserId: String
    ): Boolean {
        val latest = latestMessage(conversation.id) ?: return false
        return !latest.isSentBy(currentUserId)
    }

    private suspend fun latestMessage(conversationId: String): Message? =
        client.postgrest.from("messages")
            .select {
                filter { eq("conversation_id", conversationId) }
                order("created_at", Order.DESCENDING)
                limit(1)
            }
            .decodeList<Message>()
            .firstOrNull()

    // MARK: - Safety actions

    // MARK: - Matches and inbound likes

    /**
     * Active matches with the other person's profile, newest first.
     *
     * Blocked users and duplicate pairings are dropped, so a pair that somehow
     * matched twice shows once.
     */
    suspend fun getMatches(userId: String): List<MatchWithProfile> {
        val blocked = getBlockedUserIds(userId)

        val matches = client.postgrest.from("matches")
            .select {
                filter {
                    or {
                        eq("user1_id", userId)
                        eq("user2_id", userId)
                    }
                    eq("is_active", true)
                }
                order("matched_at", Order.DESCENDING)
            }
            .decodeList<Match>()

        val result = mutableListOf<MatchWithProfile>()
        val seen = mutableSetOf<String>()

        for (match in matches) {
            val otherId = match.otherUserId(userId) ?: continue
            if (blocked.contains(otherId.lowercase())) continue
            if (!seen.add(otherId)) continue

            val profile = runCatching { profileService.getProfile(otherId) }.getOrNull() ?: continue
            result.add(MatchWithProfile(match, profile))
        }

        return result
    }

    /**
     * People who liked this user and are still waiting on an answer.
     *
     * Anyone already matched, already swiped on, or blocked is filtered out —
     * the list is only the ones a reply would still change something for.
     */
    suspend fun getInboundLikes(userId: String): List<InboundLikeWithProfile> {
        val blocked = getBlockedUserIds(userId)

        val inbound = client.postgrest.from("swipes")
            .select {
                filter {
                    eq("swiped_id", userId)
                    isIn("action", listOf(SwipeAction.LIKE.raw, SwipeAction.SUPER_LIKE.raw))
                }
                order("created_at", Order.DESCENDING)
            }
            .decodeList<Swipe>()

        if (inbound.isEmpty()) return emptyList()

        val matches = client.postgrest.from("matches")
            .select {
                filter {
                    or {
                        eq("user1_id", userId)
                        eq("user2_id", userId)
                    }
                }
            }
            .decodeList<Match>()

        val outgoing = client.postgrest.from("swipes")
            .select { filter { eq("swiper_id", userId) } }
            .decodeList<Swipe>()

        val matchedUserIds = matches.mapNotNull { it.otherUserId(userId)?.lowercase() }.toSet()
        val actedOnUserIds = outgoing.map { it.swipedId.lowercase() }.toSet()

        val result = mutableListOf<InboundLikeWithProfile>()
        val seen = mutableSetOf<String>()

        for (swipe in inbound) {
            val swiperId = swipe.swiperId.lowercase()
            if (matchedUserIds.contains(swiperId)) continue
            if (actedOnUserIds.contains(swiperId)) continue
            if (blocked.contains(swiperId)) continue
            if (!seen.add(swiperId)) continue

            val profile = runCatching { profileService.getProfile(swipe.swiperId) }.getOrNull()
                ?: continue
            result.add(InboundLikeWithProfile(swipe, profile))
        }

        return result
    }

    /**
     * Answers an inbound like. Returns the match id when the answer created a
     * match, which only a like or super-like can do.
     *
     * Mirrors `saveSwipe` in Harvest/Services/SwipeService.swift, including its
     * duplicate-key tolerance: swiping twice is a no-op, not an error.
     */
    suspend fun saveSwipe(swiperId: String, swipedId: String, action: SwipeAction): String? {
        try {
            client.postgrest.from("swipes").insert(
                buildJsonObject {
                    put("swiper_id", swiperId)
                    put("swiped_id", swipedId)
                    put("action", action.raw)
                }
            )
        } catch (e: Exception) {
            // 23505 is the unique violation: already swiped, nothing to do.
            if (e.message?.contains("23505") != true) throw e
            return null
        }

        if (action == SwipeAction.NOPE) return null

        @Serializable
        data class MatchStatus(
            @SerialName("is_matched") val isMatched: Boolean = false,
            @SerialName("match_id") val matchId: String? = null
        )

        val status = runCatching {
            client.postgrest.rpc(
                "get_match_status",
                buildJsonObject {
                    put("user_a", swiperId)
                    put("user_b", swipedId)
                }
            ).decodeList<MatchStatus>().firstOrNull()
        }.getOrNull()

        return if (status?.isMatched == true) status.matchId else null
    }

    suspend fun unmatchUser(matchId: String) {
        client.postgrest.from("matches").update(
            buildJsonObject {
                put("is_active", false)
                put("unmatched_at", Instant.now().toString())
            }
        ) { filter { eq("id", matchId) } }
    }

    suspend fun reportUser(
        reporterId: String,
        reportedUserId: String,
        category: String,
        description: String,
        targetType: String = "profile",
        targetId: String? = null
    ) {
        client.postgrest.from("user_reports").insert(
            buildJsonObject {
                put("reporter_id", reporterId)
                put("reported_id", reportedUserId)
                put("reason", category)
                put("description", description)
                put("target_type", targetType)
                if (targetId != null) put("target_id", targetId)
            }
        )
    }

    suspend fun blockUser(
        userId: String,
        blockedUserId: String,
        reason: String = "Blocked",
        description: String = "User blocked — auto-filed for moderator review."
    ) {
        client.postgrest.from("user_blocks").insert(
            buildJsonObject {
                put("blocker_id", userId)
                put("blocked_id", blockedUserId)
            }
        )

        // Apple 1.2: blocking must also notify the developer of the content so
        // it can be reviewed within 24h. File a report alongside; a failure
        // here must not undo the block, which is why it is swallowed.
        runCatching {
            reportUser(
                reporterId = userId,
                reportedUserId = blockedUserId,
                category = reason,
                description = description
            )
        }

        // Deactivate any live match between the two.
        val matches = client.postgrest.from("matches")
            .select {
                filter {
                    or {
                        eq("user1_id", userId)
                        eq("user2_id", userId)
                    }
                    eq("is_active", true)
                }
            }
            .decodeList<Match>()

        matches.filter { it.otherUserId(userId)?.lowercase() == blockedUserId.lowercase() }
            .forEach { unmatchUser(it.id) }
    }

    /**
     * Every user id in a block relationship with [userId], either direction,
     * lowercased, so callers can exclude them from feeds instantly.
     */
    suspend fun getBlockedUserIds(userId: String): Set<String> {
        @Serializable
        data class UserBlockRow(
            @SerialName("blocker_id") val blockerId: String? = null,
            @SerialName("blocked_id") val blockedId: String? = null
        )

        val rows = client.postgrest.from("user_blocks")
            .select {
                filter {
                    or {
                        eq("blocker_id", userId)
                        eq("blocked_id", userId)
                    }
                }
            }
            .decodeList<UserBlockRow>()

        val me = userId.lowercase()
        return rows
            .flatMap { listOfNotNull(it.blockerId, it.blockedId) }
            .map { it.lowercase() }
            .filterNot { it == me }
            .toSet()
    }

    companion object {
        private const val PREVIEW_LENGTH = 100
    }
}
